File: app/src/main/kotlin/net/tigr/soulcalc/ui/viewmodel/MainViewModel.kt
/*
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2026 Albert Zenkoff
 */

package net.tigr.soulcalc.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import net.tigr.soulcalc.data.repository.SheetRepository
import net.tigr.soulcalc.domain.engine.SheetEngine
import net.tigr.soulcalc.domain.model.Line
import net.tigr.soulcalc.domain.model.Result
import net.tigr.soulcalc.domain.parser.LineClassifier
import kotlin.math.abs

/**
 * ViewModel for the main calculator screen.
 *
 * Manages UI state and coordinates with SheetEngine for calculations.
 * Handles persistence through SheetRepository with debounced auto-save.
 */
class MainViewModel(
    private val repository: SheetRepository?,
    private val engine: SheetEngine = SheetEngine()
) : ViewModel() {

    private val _uiState = MutableStateFlow(MainUiState())
    val uiState: StateFlow<MainUiState> = _uiState.asStateFlow()

    private var currentSheetId: String? = null
    private var saveJob: Job? = null
    private var focusSaveJob: Job? = null

    init {
        loadSheet()
    }

    private fun loadSheet() {
        if (repository == null) {
            updateLines(engine.evaluate(listOf("")))
            return
        }

        viewModelScope.launch {
            val sheet = repository.loadOrCreateSheet()
            currentSheetId = sheet.id
            val inputs = sheet.lines.map { it.input }
            val lines = engine.evaluate(inputs.ifEmpty { listOf("") })
            val uiLines = lines.map { it.toUiModel() }
            val restoredFocusIndex = sheet.focusedLineIndex.coerceIn(0, lines.size.coerceAtLeast(1) - 1)
            // Single atomic update for lines and focus
            _uiState.update { it.copy(
                lines = uiLines,
                focusedLineIndex = restoredFocusIndex
            )}
        }
    }

    /**
     * Handles UI events from the view.
     */
    fun onEvent(event: MainUiEvent) {
        when (event) {
            is MainUiEvent.LineTextChanged -> handleLineTextChanged(event.lineIndex, event.text)
            is MainUiEvent.NewLineRequested -> handleNewLine(event.afterLineIndex)
            is MainUiEvent.DeleteLineRequested -> handleDeleteLine(event.lineIndex)
            is MainUiEvent.CopyResult -> handleCopyResult(event.lineIndex)
            is MainUiEvent.CopyAll -> handleCopyAll()
            is MainUiEvent.ClearSheet -> handleClearSheet()
            is MainUiEvent.DollarKeyPressed -> handleDollarKey()
            is MainUiEvent.HashKeyPressed -> handleHashKey()
            is MainUiEvent.KeyPressed -> handleKeyPressed(event.key)
            is MainUiEvent.EnterKeyPressed -> handleEnterKey(event.afterLineIndex)
            is MainUiEvent.ToggleKeyboard -> handleToggleKeyboard()
            is MainUiEvent.ShowVariablePicker -> handleShowVariablePicker()
            is MainUiEvent.VariableSelected -> handleVariableSelected(event.variableName)
            is MainUiEvent.DismissVariablePicker -> handleDismissVariablePicker()
            is MainUiEvent.LineFocused -> handleLineFocused(event.lineIndex)
            is MainUiEvent.ToastShown -> handleToastShown()
            is MainUiEvent.InsertionConsumed -> handleInsertionConsumed()
            is MainUiEvent.BackspacePressed -> handleBackspacePressed()
            is MainUiEvent.BackspaceConsumed -> handleBackspaceConsumed()
            is MainUiEvent.ClearLinePressed -> handleClearLinePressed()
            is MainUiEvent.ClearLineConsumed -> handleClearLineConsumed()
        }
    }

    private fun handleLineTextChanged(lineIndex: Int, text: String) {
        val lines = engine.updateLine(lineIndex, text)
        updateLines(lines)
        scheduleSave()
    }

    private fun handleNewLine(afterLineIndex: Int) {
        val lines = engine.insertLine(afterLineIndex + 1, "")
        val uiLines = lines.map { it.toUiModel() }
        // Single atomic update for lines, focus, and keyboard state
        _uiState.update { it.copy(
            lines = uiLines,
            focusedLineIndex = afterLineIndex + 1,
            useCustomKeyboard = true
        )}
        scheduleSave()
    }

    private fun handleDeleteLine(lineIndex: Int) {
        val currentLines = engine.getLines()

        // Validate bounds
        if (lineIndex < 0 || lineIndex >= currentLines.size) return

        if (currentLines.size <= 1) {
            val lines = engine.updateLine(0, "")
            updateLines(lines)
            scheduleSave()
            return
        }

        val lines = engine.removeLine(lineIndex)
        val uiLines = lines.map { it.toUiModel() }
        // Single atomic update for lines and focus
        _uiState.update { it.copy(
            lines = uiLines,
            focusedLineIndex = (lineIndex - 1).coerceAtLeast(0)
        )}
        scheduleSave()
    }

    private fun handleCopyResult(lineIndex: Int) {
        val lines = _uiState.value.lines
        if (lineIndex in lines.indices) {
            val resultText = lines[lineIndex].resultText
            if (resultText.isNotEmpty() && !lines[lineIndex].isError) {
                _uiState.update { it.copy(toastMessage = "Copied: $resultText") }
            }
        }
    }

    private fun handleCopyAll() {
        val formatted = formatAllLines()
        if (formatted.isNotEmpty()) {
            _uiState.update { it.copy(toastMessage = "Copied all lines") }
        }
    }

    private fun handleClearSheet() {
        val lines = engine.clear()
        val uiLines = lines.map { it.toUiModel() }
        // Single atomic update for lines, focus, and keyboard state
        _uiState.update { it.copy(
            lines = uiLines,
            focusedLineIndex = 0,
            useCustomKeyboard = true
        )}
        scheduleSave()
    }

    private fun handleDollarKey() {
        // Set pending insertion for "$" and switch to system keyboard.
        // LineRow will insert at cursor position via pendingInsertion handler.
        _uiState.update { it.copy(pendingInsertion = "$", useCustomKeyboard = false) }
    }

    private fun handleHashKey() {
        // Set pending insertion for "#" and switch to system keyboard.
        // LineRow will insert at cursor position via pendingInsertion handler.
        _uiState.update { it.copy(pendingInsertion = "#", useCustomKeyboard = false) }
    }

    private fun handleEnterKey(afterLineIndex: Int) {
        // Delegate to handleNewLine - both do the same: create line and switch keyboard
        handleNewLine(afterLineIndex)
    }

    private fun handleToggleKeyboard() {
        _uiState.update { it.copy(useCustomKeyboard = !it.useCustomKeyboard) }
    }

    private fun handleShowVariablePicker() {
        val scope = engine.getScope()
        val variables = scope.variables.keys.toList().sorted()
        _uiState.update { it.copy(
            showVariablePicker = true,
            availableVariables = variables
        )}
    }

    private fun handleVariableSelected(variableName: String) {
        _uiState.update { it.copy(
            showVariablePicker = false,
            pendingInsertion = "\$$variableName"
        )}
    }

    private fun handleDismissVariablePicker() {
        _uiState.update { it.copy(showVariablePicker = false) }
    }

    private fun handleLineFocused(lineIndex: Int) {
        _uiState.update { it.copy(focusedLineIndex = lineIndex) }
        scheduleFocusSave(lineIndex)
    }

    private fun handleToastShown() {
        _uiState.update { it.copy(toastMessage = null) }
    }

    private fun handleInsertionConsumed() {
        _uiState.update { it.copy(pendingInsertion = null) }
    }

    private fun handleKeyPressed(key: String) {
        // Set pending insertion - LineRow will insert at cursor position
        _uiState.update { it.copy(pendingInsertion = key) }
    }

    private fun handleBackspacePressed() {
        // Set pending backspace - LineRow will delete at cursor position
        _uiState.update { it.copy(pendingBackspace = true) }
    }

    private fun handleBackspaceConsumed() {
        _uiState.update { it.copy(pendingBackspace = false) }
    }

    private fun handleClearLinePressed() {
        // Set pending clear line - LineRow will clear the entire line
        _uiState.update { it.copy(pendingClearLine = true) }
    }

    private fun handleClearLineConsumed() {
        _uiState.update { it.copy(pendingClearLine = false) }
    }

    private fun updateLines(lines: List<Line>) {
        val uiLines = lines.map { it.toUiModel() }
        _uiState.update { it.copy(lines = uiLines) }
    }

    private fun scheduleSave() {
        if (repository == null || currentSheetId == null) return

        // Capture state snapshot BEFORE delay to avoid race conditions
        val sheetId = currentSheetId ?: return
        val inputs = engine.getLines().map { it.input }

        saveJob?.cancel()
        saveJob = viewModelScope.launch {
            delay(SAVE_DEBOUNCE_MS)
            repository.saveLines(sheetId, inputs)
        }
    }

    private fun scheduleFocusSave(focusedLineIndex: Int) {
        if (repository == null || currentSheetId == null) return

        // Capture state snapshot BEFORE delay to avoid race conditions
        val sheetId = currentSheetId ?: return

        focusSaveJob?.cancel()
        focusSaveJob = viewModelScope.launch {
            delay(SAVE_DEBOUNCE_MS)
            repository.saveFocusedLineIndex(sheetId, focusedLineIndex)
        }
    }

    private fun Line.toUiModel(): LineUiModel {
        return LineUiModel(
            id = this.id,
            input = this.input,
            resultText = formatResult(this.result),
            isError = this.result is Result.Error,
            isComment = LineClassifier.isComment(this.input),
            isEmpty = LineClassifier.isEmpty(this.input)
        )
    }

    private fun formatResult(result: Result): String {
        return when (result) {
            is Result.Success -> formatNumber(result.value)
            is Result.Error -> formatError(result.message)
            is Result.Empty -> ""
        }
    }

    private fun formatError(message: String): String {
        return when {
            // Display-ready symbols from Evaluator - pass through as-is
            message == "∞" || message == "-∞" || message == "NaN" -> message
            // Error messages with context
            message.contains("undefined variable", ignoreCase = true) -> {
                val varName = Regex("\\$\\w+").find(message)?.value ?: ""
                "? $varName"
            }
            message.contains("invalid line reference", ignoreCase = true) -> {
                val lineRef = Regex("\\$\\d+").find(message)?.value ?: ""
                "? $lineRef"
            }
            // Evaluator error messages that include context (e.g., "? $varName")
            message.startsWith("?") -> message
            else -> "?"
        }
    }

    private fun formatNumber(value: Double): String {
        if (value.isNaN()) return "NaN"
        if (value.isInfinite()) return if (value > 0) "∞" else "-∞"

        if (value == value.toLong().toDouble() && abs(value) < 1e15) {
            return value.toLong().toString()
        }

        val formatted = "%.10g".format(value)
        // Strip trailing zeros and unnecessary decimal point
        return formatted
            .replace(Regex("(\\.\\d*?)0+$"), "$1")
            .replace(Regex("\\.$"), "")
    }

    /**
     * Formats all lines for "Copy All" functionality.
     */
    fun formatAllLines(): String {
        return _uiState.value.lines
            .filter { !it.isEmpty }
            .joinToString("\n") { line ->
                if (line.isComment) {
                    line.input
                } else if (line.resultText.isNotEmpty()) {
                    "${line.input} = ${line.resultText}"
                } else {
                    line.input
                }
            }
    }

    /**
     * Gets the result text for a specific line (for clipboard operations).
     */
    fun getResultForClipboard(lineIndex: Int): String? {
        val lines = _uiState.value.lines
        if (lineIndex !in lines.indices) return null
        val line = lines[lineIndex]
        if (line.isError || line.resultText.isEmpty()) return null
        return line.resultText
    }

    companion object {
        private const val SAVE_DEBOUNCE_MS = 500L
    }

    class Factory(private val repository: SheetRepository) : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            if (modelClass.isAssignableFrom(MainViewModel::class.java)) {
                return MainViewModel(repository) as T
            }
            throw IllegalArgumentException("Unknown ViewModel class")
        }
    }
}

File: ios/SoulCalc/UI/ViewModel/MainViewModel.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Albert Zenkoff

import SwiftUI
import Combine

// MARK: - UI Models

/// UI model for a single line in the calculator.
struct LineUiModel: Identifiable {
    let id: Int
    let input: String
    let resultText: String
    let isError: Bool
    let isComment: Bool
    let isEmpty: Bool

    init(id: Int = 0, input: String = "", resultText: String = "",
         isError: Bool = false, isComment: Bool = false, isEmpty: Bool = true) {
        self.id = id
        self.input = input
        self.resultText = resultText
        self.isError = isError
        self.isComment = isComment
        self.isEmpty = isEmpty
    }
}

/// UI state for the main calculator screen.
struct MainUiState {
    var lines: [LineUiModel] = [LineUiModel()]
    var focusedLineIndex: Int = 0
    var useCustomKeyboard: Bool = true
    var showVariablePicker: Bool = false
    var availableVariables: [String] = []
    var toastMessage: String? = nil
    var pendingInsertion: String? = nil
    var pendingBackspace: Bool = false
    var pendingClearLine: Bool = false
}

// MARK: - ViewModel

/// ViewModel for the main calculator screen.
///
/// Manages UI state and coordinates with SheetEngine for calculations.
/// Handles persistence through SheetStore with debounced auto-save.
@MainActor
class MainViewModel: ObservableObject {
    @Published var state = MainUiState()

    private let engine = SheetEngine()
    private let store: SheetStore?
    private var saveTask: Task<Void, Never>?
    private var focusSaveTask: Task<Void, Never>?

    private static let saveDebounceNs: UInt64 = 500_000_000 // 500ms

    init(store: SheetStore? = SheetStore()) {
        self.store = store
        loadSheet()
    }

    private func loadSheet() {
        guard let store = store, let data = store.load(), !data.lines.isEmpty else {
            let lines = engine.evaluate([""])
            updateLines(lines)
            return
        }

        let lines = engine.evaluate(data.lines)
        let uiLines = lines.map { $0.toUiModel() }
        let restoredFocus = max(0, min(data.focusedLineIndex, lines.count - 1))
        state.lines = uiLines
        state.focusedLineIndex = restoredFocus
    }

    // MARK: - Event Handlers

    func lineTextChanged(_ lineIndex: Int, _ text: String) {
        let lines = engine.updateLine(lineIndex, text)
        updateLines(lines)
        scheduleSave()
    }

    func newLineRequested(_ afterLineIndex: Int) {
        let lines = engine.insertLine(afterLineIndex + 1, "")
        let uiLines = lines.map { $0.toUiModel() }
        state.lines = uiLines
        state.focusedLineIndex = afterLineIndex + 1
        state.useCustomKeyboard = true
        scheduleSave()
    }

    func deleteLineRequested(_ lineIndex: Int) {
        let currentLines = engine.getLines()
        if lineIndex < 0 || lineIndex >= currentLines.count { return }

        if currentLines.count <= 1 {
            let lines = engine.updateLine(0, "")
            updateLines(lines)
            scheduleSave()
            return
        }

        let lines = engine.removeLine(lineIndex)
        let uiLines = lines.map { $0.toUiModel() }
        state.lines = uiLines
        state.focusedLineIndex = max(lineIndex - 1, 0)
        scheduleSave()
    }

    func copyResult(_ lineIndex: Int) -> String? {
        let lines = state.lines
        guard lineIndex >= 0, lineIndex < lines.count else { return nil }
        let line = lines[lineIndex]
        guard !line.isError, !line.resultText.isEmpty else { return nil }
        state.toastMessage = "Copied: \(line.resultText)"
        return line.resultText
    }

    func copyAll() -> String {
        let formatted = formatAllLines()
        if !formatted.isEmpty {
            state.toastMessage = "Copied all lines"
        }
        return formatted
    }

    func clearSheet() {
        let lines = engine.clear()
        let uiLines = lines.map { $0.toUiModel() }
        state.lines = uiLines
        state.focusedLineIndex = 0
        state.useCustomKeyboard = true
        scheduleSave()
    }

    func dollarKeyPressed() {
        state.pendingInsertion = "$"
        state.useCustomKeyboard = false
    }

    func hashKeyPressed() {
        state.pendingInsertion = "#"
        state.useCustomKeyboard = false
    }

    func keyPressed(_ key: String) {
        state.pendingInsertion = key
    }

    func enterKeyPressed() {
        newLineRequested(state.focusedLineIndex)
    }

    func toggleKeyboard() {
        state.useCustomKeyboard.toggle()
    }

    func showVariablePicker() {
        let scope = engine.getScope()
        let variables = scope.variables.keys.sorted()
        state.showVariablePicker = true
        state.availableVariables = variables
    }

    func variableSelected(_ name: String) {
        state.showVariablePicker = false
        state.pendingInsertion = "$\(name)"
    }

    func dismissVariablePicker() {
        state.showVariablePicker = false
    }

    func lineFocused(_ lineIndex: Int) {
        state.focusedLineIndex = lineIndex
        scheduleFocusSave(lineIndex)
    }

    func toastShown() {
        state.toastMessage = nil
    }

    func insertionConsumed() {
        state.pendingInsertion = nil
    }

    func backspacePressed() {
        state.pendingBackspace = true
    }

    func backspaceConsumed() {
        state.pendingBackspace = false
    }

    func clearLinePressed() {
        state.pendingClearLine = true
    }

    func clearLineConsumed() {
        state.pendingClearLine = false
    }

    // MARK: - Formatting

    func formatAllLines() -> String {
        state.lines
            .filter { !$0.isEmpty }
            .map { line in
                if line.isComment {
                    return line.input
                } else if !line.resultText.isEmpty {
                    return "\(line.input) = \(line.resultText)"
                } else {
                    return line.input
                }
            }
            .joined(separator: "\n")
    }

    // MARK: - Private

    private func updateLines(_ lines: [Line]) {
        state.lines = lines.map { $0.toUiModel() }
    }

    private func scheduleSave() {
        guard let store = store else { return }
        let inputs = engine.getLines().map { $0.input }

        saveTask?.cancel()
        saveTask = Task {
            try? await Task.sleep(nanoseconds: Self.saveDebounceNs)
            guard !Task.isCancelled else { return }
            store.saveLines(inputs)
        }
    }

    private func scheduleFocusSave(_ index: Int) {
        guard store != nil else { return }

        focusSaveTask?.cancel()
        focusSaveTask = Task {
            try? await Task.sleep(nanoseconds: Self.saveDebounceNs)
            guard !Task.isCancelled else { return }
            store?.saveFocusedLineIndex(index)
        }
    }
}

// MARK: - Line → UI Model conversion

private extension Line {
    func toUiModel() -> LineUiModel {
        LineUiModel(
            id: self.id,
            input: self.input,
            resultText: Self.formatResult(self.result),
            isError: {
                if case .error = self.result { return true }
                return false
            }(),
            isComment: LineClassifier.isComment(self.input),
            isEmpty: LineClassifier.isEmpty(self.input)
        )
    }

    static func formatResult(_ result: Result) -> String {
        switch result {
        case .success(let value):
            return formatNumber(value)
        case .error(let message):
            return formatError(message)
        case .empty:
            return ""
        }
    }

    static func formatError(_ message: String) -> String {
        if message == "∞" || message == "-∞" || message == "NaN" {
            return message
        }
        if message.hasPrefix("?") {
            return message
        }
        return "?"
    }

    static func formatNumber(_ value: Double) -> String {
        if value.isNaN { return "NaN" }
        if value.isInfinite { return value > 0 ? "∞" : "-∞" }

        // Display integers without decimal point
        if value == Double(Int64(value)) && abs(value) < 1e15 {
            return String(Int64(value))
        }

        // Format with up to 10 significant digits, strip trailing zeros
        let formatted = String(format: "%.10g", value)
        return formatted
            .replacingOccurrences(of: "(\\.\\d*?)0+$", with: "$1", options: .regularExpression)
            .replacingOccurrences(of: "\\.$", with: "", options: .regularExpression)
    }
}
